src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -257,36 +316,46 @@ public boolean isFingerprintArtifacts() {
+    public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        if (build instanceof AbstractBuild) {
+            upgradeIfNecessary(((AbstractBuild)build).getProject());
+        }
+
+        EnvVars env = build.getEnvironment(listener);
+        if (build instanceof AbstractBuild) {
+            env.overrideAll(((AbstractBuild)build).getBuildVariables()); // Add in matrix axes..
+        }
+
         PrintStream console = listener.getLogger();
         String expandedProject = project, expandedFilter = filter;
         String expandedExcludes = getExcludes();
         try {
-            EnvVars env = build.getEnvironment(listener);
-            env.overrideAll(build.getBuildVariables()); // Add in matrix axes..
             expandedProject = env.expand(project);
-            Job<?, ?> job = Jenkins.getInstance().getItem(expandedProject, build.getProject().getRootProject().getParent(), Job.class);
+            Job<?, ?> job = Jenkins.getInstance().getItem(expandedProject, getItemGroup(build), Job.class);
             if (job != null && !expandedProject.equals(project)
                 // If projectName is parameterized, need to do permission check on source project.
                 && !canReadFrom(job, build)) {
                 job = null; // Disallow access
             }
             if (job == null) {
-                console.println(Messages.CopyArtifact_MissingProject(expandedProject));
-                return false;
+                String message = Messages.CopyArtifact_MissingProject(expandedProject);
+                console.println(message);
+                throw new CopyArtifactException(message);
             }
             Run src = selector.getBuild(job, env, parameters != null ? new ParametersBuildFilter(env.expand(parameters)) : new BuildFilter(), build);
             if (src == null) {
-                console.println(Messages.CopyArtifact_MissingBuild(expandedProject));
-                return isOptional();  // Fail build unless copy is optional
-            }
-            FilePath targetDir = build.getWorkspace(), baseTargetDir = targetDir;
-            if (targetDir == null || !targetDir.exists()) {
-                console.println(Messages.CopyArtifact_MissingWorkspace()); // (see JENKINS-3330)
-                return isOptional();  // Fail build unless copy is optional
+                String message = Messages.CopyArtifact_MissingBuild(expandedProject);
+                console.println(message);
+                if (isOptional()) {
+                    // just return without an error
+                    return;
+                } else {
+                    // Fail build if copy is not optional
+                    throw new CopyArtifactException(message);
+                }
             }
+            FilePath targetDir = workspace, baseTargetDir = targetDir;
```

Comment:
Presumably only one of these is mutated, so would suffice to rename it to `workspace` and remove the redundant local var.